x-pack/filebeat/input/websocket/input.go

```diff
@@ -0,0 +1,438 @@
+		// single evaluation. Rather than incurring the cost of creating a new
+		// cel.Program for each evaluation, shadow lib.Time's now with a new
+		// value for each eval. We retain the lib.Time now global for
+		// compatibility between CEL programs developed in mito with programs
+		// run in the input.
+		"now": now,
+		root:  state,
+	})
+	if err != nil {
+		err = lib.DecoratedError{AST: ast, Err: err}
+	}
+	if e := ctx.Err(); e != nil {
+		err = e
+	}
+	if err != nil {
+		state["events"] = errorMessage(fmt.Sprintf("failed eval: %v", err))
+		clearWantMore(state)
+		return state, fmt.Errorf("failed eval: %w", err)
+	}
+
+	v, err := out.ConvertToNative(reflect.TypeOf((*structpb.Struct)(nil)))
+	if err != nil {
+		state["events"] = errorMessage(fmt.Sprintf("failed proto conversion: %v", err))
+		clearWantMore(state)
+		return state, fmt.Errorf("failed proto conversion: %w", err)
+	}
+	switch v := v.(type) {
+	case *structpb.Struct:
+		return v.AsMap(), nil
+	default:
+		// This should never happen.
+		errMsg := fmt.Sprintf("unexpected native conversion type: %T", v)
+		state["events"] = errorMessage(errMsg)
+		clearWantMore(state)
+		return state, errors.New(errMsg)
+	}
+}
+
+// now is time.Now with a modifiable time source.
+func (i input) now() time.Time {
+	if i.time == nil {
+		return time.Now()
+	}
+	return i.time()
+}
+
+// clearWantMore sets the state to not request additional work in a periodic evaluation.
+// It leaves state intact if there is no "want_more" element, and sets the element to false
+// if there is. This is necessary instead of just doing delete(state, "want_more") as
+// client CEL code may expect the want_more field to be present.
+func clearWantMore(state map[string]interface{}) {
+	if _, ok := state["want_more"]; ok {
+		state["want_more"] = false
+	}
+}
+
+func errorMessage(msg string) map[string]interface{} {
+	return map[string]interface{}{"error": map[string]interface{}{"message": msg}}
+}
+
+func wantClient(cfg config) bool {
+	switch scheme, _, _ := strings.Cut(cfg.Resource.URL.Scheme, "+"); scheme {
+	case "ws", "wss":
+		return true
+	default:
+		return false
+	}
+}
+
+func formHeader(cfg config) map[string][]string {
+	header := make(map[string][]string)
+	switch {
+	case cfg.Auth.ApiKey != nil:
+		header[cfg.Auth.ApiKey.Header] = []string{cfg.Auth.ApiKey.Value}
+	case cfg.Auth.BearerToken != "":
+		header["Authorization"] = []string{"Bearer " + cfg.Auth.BearerToken}
+	case cfg.Auth.BasicToken != "":
+		header["Authorization"] = []string{"Basic " + cfg.Auth.BasicToken}
+	}
+	return header
+}
+
+func connectWebSocketWithRetry(log *logp.Logger, url string, config retryConfig) (*websocket.Conn, error) {
+	var conn *websocket.Conn
+	var err error
+
+	for attempt := 1; attempt <= *config.MaxAttempts; attempt++ {
+		conn, _, err = websocket.DefaultDialer.Dial(url, nil)
+		if err == nil {
+			return conn, nil
+		}
+
+		log.Debugw("Attempt %d: WebSocket connection failed. Retrying...\n", attempt)
```

Comment:
~All the logs here are written to the *w logger (as in this case), so there should be no formatting verbs. Also, lowercase and the retrying is not true for the last iteration, so just drop it.
```suggestion
		log.Debugw("websocket connection failed", "attempt", attempt, "max_attempts", *config.MaxAttempts)
```